xmppserver/src/main/java/org/jivesoftware/util/SAXReaderUtil.java

```diff
@@ -0,0 +1,287 @@
+     *
+     * Equivalent to calling <tt>readDocument(input).getRootElement()</tt>
+     *
+     * @param text The data to be parsed as an XML Document.
+     * @return an XML Document root element.
+     */
+    public static Element readRootElement(@Nonnull final String text) throws ExecutionException, InterruptedException {
+        return readDocument(text).getRootElement();
+    }
+
+    /**
+     * Parses an XML Document from the content provided by a reader.
+     *
+     * The XML parsing is delegated to a limited set of parsers. When a parser is not immediately available, this method
+     * will block until one becomes available. The availability of parsers can be optimized by using the SystemProperties
+     * in this class.
+     *
+     * @param reader The data to be parsed as an XML Document.
+     * @return an XML Document.
+     */
+    public static Document readDocument(@Nonnull final Reader reader) throws ExecutionException, InterruptedException {
+        return parserService.submit(new ParserTask(reader)).get();
+    }
+
+    /**
+     * Parses an XML Document from the content provided by a reader and returns the root element of the document.
+     *
+     * Equivalent to calling <tt>readDocument(input).getRootElement()</tt>
+     *
+     * @param reader The data to be parsed as an XML Document.
+     * @return an XML Document root element.
+     */
+    public static Element readRootElement(@Nonnull final Reader reader) throws ExecutionException, InterruptedException {
+        return readDocument(reader).getRootElement();
+    }
+
+    /**
+     * Parses an XML Document from the content provided by a file.
+     *
+     * The XML parsing is delegated to a limited set of parsers. When a parser is not immediately available, this method
+     * will block until one becomes available. The availability of parsers can be optimized by using the SystemProperties
+     * in this class.
+     *
+     * @param file The data to be parsed as an XML Document.
+     * @return an XML Document.
+     */
+    public static Document readDocument(@Nonnull final File file) throws ExecutionException, InterruptedException {
+        return parserService.submit(new ParserTask(file)).get();
+    }
+
+    /**
+     * Parses an XML Document from the content provided by a file and returns the root element of the document.
+     *
+     * Equivalent to calling <tt>readDocument(input).getRootElement()</tt>
+     *
+     * @param file The data to be parsed as an XML Document.
+     * @return an XML Document root element.
+     */
+    public static Element readRootElement(@Nonnull final File file) throws ExecutionException, InterruptedException {
+        return readDocument(file).getRootElement();
+    }
+
+    /**
+     * Constructs a new SAXReader instance, configured for XMPP parsing.
+     *
+     * @return A new SAXReader instance.
+     */
+    private static SAXReader constructNewReader() throws SAXException
```

Comment:
I'd question of the overhead of setting up a new SAXReader is really worth the added complexity of thread pools, etc. Is it so great?

Reply:
Definitely, and especially visible during server startup (when a lot of text pulled from the database is being parsed). Take, for example, [Daniel's comment](https://github.com/igniterealtime/Openfire/pull/1844#issuecomment-861205776), where re-using a SAXReader instead of creating a new instance for each item of data to be parsed causes the server startup time to be reduced from 30-15 minutes to 3 minutes.